Cannot unseal vault with wrong plugin checksum

This issue is probably related to #3241, but covers another failure mode.
When having a vault plugin mounted, if the checksum set in `sys/plugins/catalog/stuff` is wrong, vault cannot be unsealed, and thus, the offending mount cannot be removed, essentially bricking the vault database. The unsealing fails with the following log: 
```
2017/11/20 15:53:41.048250 [INFO ] core: vault is unsealed
2017/11/20 15:53:41.050187 [INFO ] core: post-unseal setup starting
2017/11/20 15:53:41.052123 [INFO ] core: loaded wrapping token key
2017/11/20 15:53:41.052161 [INFO ] core: successfully setup plugin catalog: plugin-directory=/vault/plugins
2017/11/20 15:53:41.055848 [INFO ] core: successfully mounted backend: type=generic path=secret/
2017/11/20 15:53:41.056002 [INFO ] core: successfully mounted backend: type=system path=sys/
2017/11/20 15:53:41.056228 [INFO ] core: successfully mounted backend: type=identity path=identity/
2017/11/20 15:53:41.111671 [ERROR] core: failed to create mount entry: path=somepath/ error=checksums did not match
2017/11/20 15:53:41.111731 [INFO ] core: pre-seal teardown starting
2017/11/20 15:53:41.111783 [INFO ] core: cluster listeners not running
2017/11/20 15:53:41.111869 [INFO ] core: pre-seal teardown complete
2017/11/20 15:53:41.111916 [ERROR] core: post-unseal setup failed: error=failed to setup mount table
2017/11/20 15:53:41.111941 [WARN ] core: vault is sealed
```

**Expected Behavior:**
Vault should unseal with a warnign and leave the offending mount inacessible. It should be possible to update the checksum and/or remove the mount.

The same happens when deleting the plugin file, it also complains that it cannot verify the checksum:
```
2017/11/22 14:27:25.057578 [ERROR] core: failed to create mount entry: path=test/ error=error verifying checksum: open /some/path/to/some/plugin: The system cannot find the file specified.
```

We actually partially addressed the problem for the unseal scenario too (in https://github.com/hashicorp/vault/pull/3255/commits/a7dd937fd5a87d8891b4db2c6de9686e8d9f2c3c), but _only_ for the case where the plugin is removed from the plugin catalog. If the plugin is removed from the catalog before sealing, during unseal the mount entry will be kept in place, but the backend will not be initialized, and that's where I remember why your wrote your last comment on 3266 (that maybe we should just entirely remove the entry).

```
2017/11/22 12:12:07.151058 [INFO ] core: vault is unsealed
2017/11/22 12:12:07.151272 [INFO ] core: post-unseal setup starting
2017/11/22 12:12:07.151301 [TRACE] core: clearing forwarding clients
2017/11/22 12:12:07.151314 [TRACE] core: done clearing forwarding clients
2017/11/22 12:12:07.151718 [INFO ] core: loaded wrapping token key
2017/11/22 12:12:07.151755 [INFO ] core: successfully setup plugin catalog: plugin-directory=/Users/cleung/code/tmp/vault_plugins
2017/11/22 12:12:07.152563 [INFO ] core: successfully mounted backend: type=kv path=secret/
2017/11/22 12:12:07.152928 [INFO ] core: successfully mounted backend: type=system path=sys/
2017/11/22 12:12:07.153382 [INFO ] core: successfully mounted backend: type=identity path=identity/
2017/11/22 12:12:07.153525 [ERROR] core: failed to create mount entry: path=mock/ error=plugin not found in the catalog: mock-plugin
2017/11/22 12:12:07.153575 [INFO ] core: successfully mounted backend: type=plugin path=mock/
2017/11/22 12:12:07.153633 [INFO ] core: successfully mounted backend: type=cubbyhole path=cubbyhole/
```

--- a/vault/auth.go
+++ b/vault/auth.go
@@ -5,7 +5,6 @@ import (
 	"fmt"
 	"strings"
 
-	"github.com/hashicorp/errwrap"
 	"github.com/hashicorp/go-uuid"
 	"github.com/hashicorp/vault/helper/consts"
 	"github.com/hashicorp/vault/helper/jsonutil"
@@ -460,10 +459,11 @@ func (c *Core) setupCredentials() error {
 		backend, err = c.newCredentialBackend(entry.Type, sysView, view, conf)
 		if err != nil {
 			c.logger.Error("core: failed to create credential entry", "path", entry.Path, "error", err)
-			if errwrap.Contains(err, ErrPluginNotFound.Error()) && entry.Type == "plugin" {
-				// If we encounter an error instantiating the backend due to it being missing from the catalog,
-				// skip backend initialization but register the entry to the mount table to preserve storage
-				// and path.
+			if entry.Type == "plugin" {
+				// If we encounter an error instantiating the backend due to an error,
+				// skip backend initialization but register the entry to the mount table
+				// to preserve storage and path.
+				c.logger.Warn("core: skipping plugin-based credential entry", "path", entry.Path)
 				goto ROUTER_MOUNT
 			}
 			return errLoadAuthFailed

--- a/vault/logical_system_integ_test.go
+++ b/vault/logical_system_integ_test.go
@@ -249,15 +249,12 @@ func testPlugin_continueOnError(t *testing.T, btype logical.BackendType, mismatc
 	}
 
 	// Trigger a sha256 mistmatch or missing plugin error
-	var reqData map[string]interface{}
 	if mismatch {
-		reqData = map[string]interface{}{
+		req = logical.TestRequest(t, logical.UpdateOperation, "sys/plugins/catalog/mock-plugin")
+		req.Data = map[string]interface{}{
 			"sha256":  "d17bd7334758e53e6fbab15745d2520765c06e296f2ce8e25b7919effa0ac216",
 			"command": filepath.Base(command),
 		}
-
-		req = logical.TestRequest(t, logical.UpdateOperation, "sys/plugins/catalog/mock-plugin")
-		req.Data = reqData
 		req.ClientToken = core.Client.Token()
 		resp, err = core.HandleRequest(req)
 		if err != nil || (resp != nil && resp.IsError()) {

--- a/vault/logical_system_integ_test.go
+++ b/vault/logical_system_integ_test.go
@@ -2,7 +2,9 @@ package vault_test
 
 import (
 	"fmt"
+	"io/ioutil"
 	"os"
+	"path/filepath"
 	"testing"
 	"time"
 
@@ -178,14 +180,13 @@ func testPlugin_CatalogRemoved(t *testing.T, btype logical.BackendType, testMoun
 	}
 
 	if testMount {
-		// Add plugin back to the catalog
-		vault.TestAddTestPlugin(t, core.Core, "mock-plugin", "TestBackend_PluginMainLogical")
-
 		// Mount the plugin at the same path after plugin is re-added to the catalog
 		// and expect an error due to existing path.
 		var err error
 		switch btype {
 		case logical.TypeLogical:
+			// Add plugin back to the catalog
+			vault.TestAddTestPlugin(t, core.Core, "mock-plugin", "TestBackend_PluginMainLogical")
 			_, err = core.Client.Logical().Write("sys/mounts/mock-0", map[string]interface{}{
 				"type": "plugin",
 				"config": map[string]interface{}{
@@ -193,6 +194,8 @@ func testPlugin_CatalogRemoved(t *testing.T, btype logical.BackendType, testMoun
 				},
 			})
 		case logical.TypeCredential:
+			// Add plugin back to the catalog
+			vault.TestAddTestPlugin(t, core.Core, "mock-plugin", "TestBackend_PluginMainCredentials")
 			_, err = core.Client.Logical().Write("sys/auth/mock-0", map[string]interface{}{
 				"type":        "plugin",
 				"plugin_name": "mock-plugin",
@@ -201,6 +204,132 @@ func testPlugin_CatalogRemoved(t *testing.T, btype logical.BackendType, testMoun
 		if err == nil {
 			t.Fatal("expected error when mounting on existing path")
 		}
+	}
+}
+
+func TestSystemBackend_Plugin_continueOnError(t *testing.T) {
+	t.Run("secret", func(t *testing.T) {
+		t.Run("sha256_mismatch", func(t *testing.T) {
+			testPlugin_continueOnError(t, logical.TypeLogical, true)
+		})
+
+		t.Run("missing_plugin", func(t *testing.T) {
+			testPlugin_continueOnError(t, logical.TypeLogical, false)
+		})
+	})
+
+	t.Run("auth", func(t *testing.T) {
+		t.Run("sha256_mismatch", func(t *testing.T) {
+			testPlugin_continueOnError(t, logical.TypeCredential, true)
+		})
+
+		t.Run("missing_plugin", func(t *testing.T) {
+			testPlugin_continueOnError(t, logical.TypeCredential, false)
+		})
+	})
+}
+
+func testPlugin_continueOnError(t *testing.T, btype logical.BackendType, mismatch bool) {
+	cluster := testSystemBackendMock(t, 1, 1, btype)
+	defer cluster.Cleanup()
+
+	core := cluster.Cores[0]
+
+	// Get the registered plugin
+	req := logical.TestRequest(t, logical.ReadOperation, "sys/plugins/catalog/mock-plugin")
+	req.ClientToken = core.Client.Token()
+	resp, err := core.HandleRequest(req)
+	if err != nil || resp == nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("err:%v resp:%#v", err, resp)
+	}
+
+	command, ok := resp.Data["command"].(string)
+	if !ok || command == "" {
+		t.Fatal("invalid command")
+	}
+
+	// Trigger a sha256 mistmatch or missing plugin error
+	var reqData map[string]interface{}
+	if mismatch {
+		reqData = map[string]interface{}{
+			"sha256":  "d17bd7334758e53e6fbab15745d2520765c06e296f2ce8e25b7919effa0ac216",
+			"command": filepath.Base(command),
+		}
+
+		req = logical.TestRequest(t, logical.UpdateOperation, "sys/plugins/catalog/mock-plugin")
+		req.Data = reqData
+		req.ClientToken = core.Client.Token()
+		resp, err = core.HandleRequest(req)
+		if err != nil || (resp != nil && resp.IsError()) {
+			t.Fatalf("err:%v resp:%#v", err, resp)
+		}
+	} else {
+		err := os.Remove(filepath.Join(cluster.TempDir, filepath.Base(command)))
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Seal the cluster
+	cluster.EnsureCoresSealed(t)
+
+	// Unseal the cluster
+	barrierKeys := cluster.BarrierKeys
+	for _, core := range cluster.Cores {
+		for _, key := range barrierKeys {
+			_, err := core.Unseal(vault.TestKeyCopy(key))
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+		sealed, err := core.Sealed()
+		if err != nil {
+			t.Fatalf("err checking seal status: %s", err)
+		}
+		if sealed {
+			t.Fatal("should not be sealed")
+		}
+		// Wait for active so post-unseal takes place
+		// If it fails, it means unseal process failed
+		vault.TestWaitActive(t, core.Core)
+	}
+
+	// Re-add the plugin to the catalog
+	switch btype {
+	case logical.TypeLogical:
+		vault.TestAddTestPluginTempDir(t, core.Core, "mock-plugin", "TestBackend_PluginMainLogical", cluster.TempDir)
+	case logical.TypeCredential:
+		vault.TestAddTestPluginTempDir(t, core.Core, "mock-plugin", "TestBackend_PluginMainCredentials", cluster.TempDir)
+	}
+
+	// Reload the plugin
+	req = logical.TestRequest(t, logical.UpdateOperation, "sys/plugins/reload/backend")
+	req.Data = map[string]interface{}{
+		"plugin": "mock-plugin",
+	}
+	req.ClientToken = core.Client.Token()
+	resp, err = core.HandleRequest(req)
+	if err != nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("err:%v resp:%#v", err, resp)
+	}
+
+	// Make a request to lazy load the plugin
+	var reqPath string
+	switch btype {
+	case logical.TypeLogical:
+		reqPath = "mock-0/internal"
+	case logical.TypeCredential:
+		reqPath = "auth/mock-0/internal"
+	}
+
+	req = logical.TestRequest(t, logical.ReadOperation, reqPath)
+	req.ClientToken = core.Client.Token()
+	resp, err = core.HandleRequest(req)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if resp == nil {
+		t.Fatalf("bad: response should not be nil")
 	}
 }
 
@@ -332,7 +461,10 @@ func testSystemBackend_PluginReload(t *testing.T, reqData map[string]interface{}
 }
 
 // testSystemBackendMock returns a systemBackend with the desired number
-// of mounted mock plugin backends
+// of mounted mock plugin backends. numMounts alternates between different
+// ways of providing the plugin_name.
+//
+// The mounts are mounted at sys/mounts/mock-[numMounts] or sys/auth/mock-[numMounts]
 func testSystemBackendMock(t *testing.T, numCores, numMounts int, backendType logical.BackendType) *vault.TestCluster {
 	coreConfig := &vault.CoreConfig{
 		LogicalBackends: map[string]logical.Factory{
@@ -343,10 +475,17 @@ func testSystemBackendMock(t *testing.T, numCores, numMounts int, backendType lo
 		},
 	}
 
+	// Create a tempdir, cluster.Cleanup will clean up this directory
+	tempDir, err := ioutil.TempDir("", "vault-test-cluster")
+	if err != nil {
+		t.Fatal(err)
+	}
+
 	cluster := vault.NewTestCluster(t, coreConfig, &vault.TestClusterOptions{
 		HandlerFunc:        vaulthttp.Handler,
 		KeepStandbysSealed: true,
 		NumCores:           numCores,
+		TempDir:            tempDir,
 	})
 	cluster.Start()
 
@@ -358,7 +497,7 @@ func testSystemBackendMock(t *testing.T, numCores, numMounts int, backendType lo
 
 	switch backendType {
 	case logical.TypeLogical:
-		vault.TestAddTestPlugin(t, core.Core, "mock-plugin", "TestBackend_PluginMainLogical")
+		vault.TestAddTestPluginTempDir(t, core.Core, "mock-plugin", "TestBackend_PluginMainLogical", tempDir)
 		for i := 0; i < numMounts; i++ {
 			// Alternate input styles for plugin_name on every other mount
 			options := map[string]interface{}{
@@ -380,7 +519,7 @@ func testSystemBackendMock(t *testing.T, numCores, numMounts int, backendType lo
 			}
 		}
 	case logical.TypeCredential:
-		vault.TestAddTestPlugin(t, core.Core, "mock-plugin", "TestBackend_PluginMainCredentials")
+		vault.TestAddTestPluginTempDir(t, core.Core, "mock-plugin", "TestBackend_PluginMainCredentials", tempDir)
 		for i := 0; i < numMounts; i++ {
 			// Alternate input styles for plugin_name on every other mount
 			options := map[string]interface{}{

--- a/vault/mount.go
+++ b/vault/mount.go
@@ -7,7 +7,6 @@ import (
 	"strings"
 	"time"
 
-	"github.com/hashicorp/errwrap"
 	"github.com/hashicorp/go-uuid"
 	"github.com/hashicorp/vault/helper/consts"
 	"github.com/hashicorp/vault/helper/jsonutil"
@@ -753,10 +752,11 @@ func (c *Core) setupMounts() error {
 		backend, err = c.newLogicalBackend(entry.Type, sysView, view, conf)
 		if err != nil {
 			c.logger.Error("core: failed to create mount entry", "path", entry.Path, "error", err)
-			if errwrap.Contains(err, ErrPluginNotFound.Error()) && entry.Type == "plugin" {
-				// If we encounter an error instantiating the backend due to it being missing from the catalog,
-				// skip backend initialization but register the entry to the mount table to preserve storage
-				// and path.
+			if entry.Type == "plugin" {
+				// If we encounter an error instantiating the backend due to an error,
+				// skip backend initialization but register the entry to the mount table
+				// to preserve storage and path.
+				c.logger.Warn("core: skipping plugin-based mount entry", "path", entry.Path)
 				goto ROUTER_MOUNT
 			}
 			return errLoadMountsFailed

--- a/vault/plugin_reload.go
+++ b/vault/plugin_reload.go
@@ -85,9 +85,13 @@ func (c *Core) reloadPluginCommon(entry *MountEntry, isAuth bool) error {
 		return nil
 	}
 
-	// Call backend's Cleanup routine
 	re := raw.(*routeEntry)
-	re.backend.Cleanup()
+
+	// Only call Cleanup if backend is initialized
+	if re.backend != nil {
+		// Call backend's Cleanup routine
+		re.backend.Cleanup()
+	}
 
 	view := re.storageView
 

--- a/vault/plugin_reload.go
+++ b/vault/plugin_reload.go
@@ -79,6 +79,10 @@ func (c *Core) reloadMatchingPlugin(pluginName string) error {
 func (c *Core) reloadPluginCommon(entry *MountEntry, isAuth bool) error {
 	path := entry.Path
 
+	if isAuth {
+		path = credentialRoutePrefix + path
+	}
+
 	// Fast-path out if the backend doesn't exist
 	raw, ok := c.router.root.Get(path)
 	if !ok {

--- a/vault/testing.go
+++ b/vault/testing.go
@@ -379,6 +379,8 @@ func TestDynamicSystemView(c *Core) *dynamicSystemView {
 	return &dynamicSystemView{c, me}
 }
 
+// TestAddTestPlugin registers the testFunc as part of the plugin command to the
+// plugin catalog.
 func TestAddTestPlugin(t testing.T, c *Core, name, testFunc string) {
 	file, err := os.Open(os.Args[0])
 	if err != nil {
@@ -405,6 +407,65 @@ func TestAddTestPlugin(t testing.T, c *Core, name, testFunc string) {
 	// Set core's plugin directory and plugin catalog directory
 	c.pluginDirectory = directoryPath
 	c.pluginCatalog.directory = directoryPath
+
+	command := fmt.Sprintf("%s --test.run=%s", filepath.Base(os.Args[0]), testFunc)
+	err = c.pluginCatalog.Set(name, command, sum)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestAddTestPluginTempDir(t testing.T, c *Core, name, testFunc, tempDir string) {
+	file, err := os.Open(os.Args[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer file.Close()
+
+	fi, err := file.Stat()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Copy over the file to the temp dir
+	dst := filepath.Join(tempDir, filepath.Base(os.Args[0]))
+	out, err := os.OpenFile(dst, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fi.Mode())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer out.Close()
+
+	if _, err = io.Copy(out, file); err != nil {
+		t.Fatal(err)
+	}
+	err = out.Sync()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fullPath, err := filepath.EvalSymlinks(tempDir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	reader, err := os.Open(filepath.Join(fullPath, filepath.Base(os.Args[0])))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Find out the sha256
+	hash := sha256.New()
+
+	_, err = io.Copy(hash, reader)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sum := hash.Sum(nil)
+
+	// Set core's plugin directory and plugin catalog directory
+	c.pluginDirectory = fullPath
+	c.pluginCatalog.directory = fullPath
 
 	command := fmt.Sprintf("%s --test.run=%s", filepath.Base(os.Args[0]), testFunc)
 	err = c.pluginCatalog.Set(name, command, sum)

